src/frollR.c

```diff
@@ -0,0 +1,209 @@
+      error("n must be integer");
+    }
+    
+    nk = length(k);
+    R_len_t i=0;                                                // check that all window values positive
+    while (i < nk && INTEGER(ik)[i] > 0) i++;
+    if (i != nk)
+      error("n must be positive integer values (> 0)");
+  } else if (badaptive) {                                       // validating n input for adaptive=TRUE
+    if (isVectorAtomic(k)) {                                    // if not-list then wrap into list
+      kl = PROTECT(allocVector(VECSXP, 1)); protecti++;
+      if (isInteger(k)) {                                       // check that k is integer vector
+        SET_VECTOR_ELT(kl, 0, k);
+      } else if (isReal(k)) {                                   // if n is double then convert to integer
+        SET_VECTOR_ELT(kl, 0, coerceVector(k, INTSXP));
+      } else {
+        error("n must be integer vector or list of integer vectors");
+      }
+      nk = 1;
+    } else {
+      nk = length(k);
+      kl = PROTECT(allocVector(VECSXP, nk)); protecti++;
+      for (R_len_t i=0; i<nk; i++) {
+        if (isInteger(VECTOR_ELT(k, i))) {
+          SET_VECTOR_ELT(kl, i, VECTOR_ELT(k, i));
+        } else if (isReal(VECTOR_ELT(k, i))) {                  // coerce double types to integer
+          SET_VECTOR_ELT(kl, i, coerceVector(VECTOR_ELT(k, i), INTSXP));
+        } else {
+          error("n must be integer vector or list of integer vectors");
+        }
+      }
+    }
+  }
+  int* ikl[nk];                                                 // pointers to adaptive window width
+  if (badaptive)
+    for (int j=0; j<nk; j++) ikl[j] = INTEGER(VECTOR_ELT(kl, j));
+  
+  if (!isLogical(narm) || length(narm)!=1 || LOGICAL(narm)[0]==NA_LOGICAL)
+    error("na.rm must be TRUE or FALSE");
+
+  if (!isLogical(hasna) || length(hasna)!=1)
+    error("hasNA must be TRUE, FALSE or NA");
+  if (LOGICAL(hasna)[0]==FALSE && LOGICAL(narm)[0])
+    error("using hasNA FALSE and na.rm TRUE does not make sense, if you know there are NA values use hasNA TRUE, otherwise leave it as default NA");
+
+  int ialign;                                                   // decode align to integer
+  if (!strcmp(CHAR(STRING_ELT(align, 0)), "right")) ialign = 1;
+  else if (!strcmp(CHAR(STRING_ELT(align, 0)), "center")) ialign = 0;
+  else if (!strcmp(CHAR(STRING_ELT(align, 0)), "left")) ialign = -1;
+  else error("Internal error: invalid align argument in rolling function, should have been caught before. please report to data.table issue tracker.");
+
+  if (badaptive && ialign!=1)
+    error("using adaptive TRUE and align argument different than 'right' is not implemented");
+  
+  SEXP ans;
+  ans = PROTECT(allocVector(VECSXP, nk * nx)); protecti++;      // allocate list to keep results
+  double* dans[nx*nk];                                          // pointers to answer columns
+  double* dx[nx];                                               // pointers to source columns
```

Comment:
@mattdowle are those double pointers safe? in other place I got segfault with such initialization. Should these be changed to `R_alloc`?